src/k8s_upgrade.py

```diff
@@ -0,0 +1,517 @@
+            self.charm.status.set_and_share_status(
+                self._upgrade.get_unit_juju_status() or ActiveStatus()
+            )
+
+    def _on_upgrade_peer_relation_created(self, _) -> None:
+        """First time the relation is created, we save the revisions."""
+        if self.charm.unit.is_leader():
+            self._upgrade.set_versions_in_app_databag()
+
+    def _on_resume_upgrade_action(self, event: ActionEvent) -> None:
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader {RESUME_ACTION_NAME}`)"
+            logger.debug(f"Resume refresh failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Resume refresh failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
+
+    def step_down_primary_and_wait_reelection(self) -> None:
+        """Steps down the current primary and waits for a new one to be elected."""
+        if len(self.charm.mongodb_config.hosts) < 2:
+            logger.warning(
+                "No secondaries to become primary - upgrading primary without electing a new one, expect downtime."
+            )
+            return
+
+        old_primary = self.charm.primary
+        with MongoDBConnection(self.charm.mongodb_config) as mongod:
+            mongod.step_down_primary()
+
+        for attempt in Retrying(stop=stop_after_attempt(15), wait=wait_fixed(1), reraise=True):
+            with attempt:
+                new_primary = self.charm.primary
+                if new_primary == old_primary:
+                    raise FailedToElectNewPrimaryError()
+
+    def run_post_app_upgrade_task(self, event: EventBase):
+        """Runs the post upgrade check to verify that the cluster is healthy.
+
+        By deferring before setting unit state to HEALTHY, the user will either:
+            1. have to wait for the unit to resolve itself.
+            2. have to run the force-upgrade action (to upgrade the next unit).
+        """
+        logger.debug(
+            "Running post refresh checks to verify the deployment is not broken after refresh."
+        )
+        self.run_post_upgrade_checks(event, finished_whole_cluster=False)
+
+        if self._upgrade.unit_state != UnitState.HEALTHY:
+            logger.info(
+                f"Unit state is not healhy but {self._upgrade.unit_state}, not continuing post-refresh checks."
+            )
+            return
+
+        # Leader of config-server must wait for all shards to be upgraded before finalising the
+        # upgrade.
+        if not self.charm.unit.is_leader() or not self.charm.is_role(Config.Role.CONFIG_SERVER):
+            logger.debug("Post refresh check is completed.")
+            self.charm.status.process_statuses()
```

Comment:
why do we call this here? is it related to your status related fix? If so can you add a comment to the code explaining the need for it, so it can be remembered in the code?

Reply:
It's not mandatory, I can remove it.
It was just to get up to date statuses after refresh :)

Reply:
Testing

Reply:
This had introduced a regression, fixing it!

Reply:
105aa6db